src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,388 @@
+
+    // Try to execute the requested method.
+    try {
+      return ctx.run(
+          [this, methodName=kj::mv(methodName), serializedArgs=kj::mv(serializedArgs),
+           entrypointName=entrypointName, callContext] (Worker::Lock& lock) mutable ->
+              kj::Promise<void> {
+
+        jsg::Lock& js = lock;
+        if (!FeatureFlags::get(js).getJsRpc()) {
+          // JS RPC is not enabled on the server side, we cannot call any methods.
+          auto rpcDisabled =
+              kj::str("The receiving Worker does not allow its methods to be called over RPC."_kj);
+          auto serialized = serializeV8(js, js.error(rpcDisabled));
+          auto builder = callContext.initResults(
+              capnp::MessageSize { serialized.size() / 8 + 8, 0 }).initException();
+          builder.setV8Serialized(kj::mv(serialized));
+          return kj::READY_NOW;
+        }
+
+        // We will try to get the function, if we can't we'll throw an error to the client.
+        auto maybeFn = tryGetFn(lock, entrypointName, ctx, methodName);
+        KJ_IF_SOME(e, maybeFn.tryGet<kj::String>()) {
+          auto serialized = serializeV8(js, js.error(e));
+          auto builder = callContext.initResults(
+              capnp::MessageSize { serialized.size() / 8 + 8, 0 }).initException();
+          builder.setV8Serialized(kj::mv(serialized));
+          return kj::READY_NOW;
         }
-    ));
+
+        auto fn = KJ_REQUIRE_NONNULL(maybeFn.tryGet<v8::Local<v8::Function>>());
+
+        // We have a function, so let's call it and serialize the result for RPC.
+        // We account for both the JS method throwing and our serialization attempt failing.
+        // If the function returns a promise we will wait for the promise to finish so we can
+        // serialize the result.
+        // TODO(now): Do we need to use awaitJs here? `call()` has to return a kj::Promise, so maybe
+        // we can return some other promise that resolves when this jsg::Promise does?
+        return ctx.awaitJs(js, callFuncAndSerialize(js, fn, serializedArgs).then(js,
+            [&](jsg::Lock& js, SerializedResult result) {
+          KJ_SWITCH_ONEOF(result) {
+            KJ_CASE_ONEOF(ok, Returned) {
+              auto builder = callContext.initResults(capnp::MessageSize {
+                  ok.serializedValue.size() / 8 + 8, 0 });
+              builder.initValue().setV8Serialized(kj::mv(ok.serializedValue));
+            }
+            KJ_CASE_ONEOF(err, ThrewException) {
+              auto builder = callContext.initResults(capnp::MessageSize {
+                  err.serializedException.size() / 8 + 8, 0 });
+              builder.initException().setV8Serialized(kj::mv(err.serializedException));
+            }
+          }
+        }));
+      }).then([this]() {
+        // We want to fulfill the callPromise so customEvent can continue executing.
+        callFulfiller->fulfill();
+      });
```

Comment:
Just want to confirm that a strong reference to `this` is attached to the `kj::Promise` we're returning correct? So that the `callFulfiller` reference is still valid by the time the `then` is triggered?